08-compiler-features.tex

```diff
@@ -142,81 +142,347 @@ \section{Dead Code Elimination}
 \section{Completion}
 \label{cr-completion}
 
-Because of the high number of Haxe features and its powerful type inference, \href{http://haxe.org/documentation/introduction/editors-and-ides.html}{IDE and code editors} cannot easily handle completion by parsing the Haxe files. Haxe provides completion support for IDEs directly built in the compiler thanks to the \ic{--display} commandline parameter.
+\subsection{Overview}
+\label{cr-completion-overview}
 
-Let's have a look at the following example:
+The rich \tref{type system}{type-system} of the Haxe Compiler makes it difficult for IDEs and editors to provide accurate completion information. Between \tref{type inference}{type-system-type-inference} and \tref{macros}{macro} it would require a substantial amount of work to replicate the required processing. This is why the Haxe Compiler comes with a built-in completion mode for third-party software to use.
+
+All completion is triggered using the \ic{--display file@position[@mode]} compiler argument. The requires arguments are:
+
+\begin{description}
+	\item[file:] The file to check for completion. This must be an absolute or relative path to a .hx file. It does not respect any class paths or libraries.
+	\item[position:] The byte position (not character position) of where to check for completion in the given file.
+	\item[mode:] The completion mode to use (see below).
+\end{description}
+
+We will look into the following completion modes in detail:
+
+\begin{description}
+	\item[\tref{Field access}{cr-completion-field-access}:] Provides a list of fields that can be accessed on a given type.
+	\item[\tref{Call argument}{cr-completion-call-argument}:] Reports the type of the function which is currently being called.
+	\item[\tref{Type path}{cr-completion-type-path}:] Lists sub-packages, sub-types and static fields.
+	\item[\tref{Usage}{cr-completion-usage}:] Lists all occurrences of a given type, field or variable in all compiled files. (mode: \ic{"usage"})
+	\item[\tref{Position}{cr-completion-position}:] Reports the position of where a given type, field or variable is defined. (mode: \ic{"position"})
+	\item[\tref{Top-level}{cr-completion-top-level}:] Lists all identifiers which are available at a given position. (mode: \ic{"toplevel"})
+\end{description}
+
+Due to Haxe being a very fast compiler it is often sufficient to rely on the normal compiler invocation for completion. For bigger projects Haxe provides a \tref{server mode}{cr-completion-server} which ensures that only those files are re-compiled that actually changed or had any of their dependencies changes.
+
```

Comment:
"very fast compiler, it is..." Consider: "Because Haxe is a very fast compiler" or even reversing into "it is often.....because Hxe is a very..."
"compiled that have actually been changed or had any of their dependencies changed" or something like this. Its not completely clear.